### DISCUSSION TITLE
Quick steps for creating a layout of app windows

### DISCUSSION THREAD

**Discussion by rxhanson** [rxhanson/RectanglePro-Community#111] [Show and tell]
There are a few ways to create a window layout, but this is probably the quickest.

1. First, select Save Current Layout... from the menu bar menu

<img width="287" alt="Untitled" src="https://user-images.githubusercontent.com/13651296/190489784-faf27ed4-c0cd-4cf2-befa-d7b74e63a8f5.png">

This will grab the coordinates and sizes of all the open windows and load them into an Applications group, and the group will automatically be displayed here.

![image](https://github.com/rxhanson/RectanglePro-Community/assets/13651296/ba648ee7-4ec5-465e-be52-ee5163775bcc)

2. It's likely that you'd want to record a keyboard shortcut here, so you can execute it and apply the group you just made.

Each of these window presets can be modified to your liking. Keep in mind that you can enter values for size and position as values less than 1 to get a percentage of the screen, and greater than one for pixel sizes.

Application groups can also be triggered by the Window Throw.

Also note that you can drag and drop window configurations for apps between app groups, so you could make layouts for each display and move them all into one group if desired.

**Comment by jschuur** [comment]
Can you pick a particular desktop on a display for an app?

I have a vertical side monitor and usually use 3-4 desktops where 2 apps each go on a particular desktop at 2/3 and 1/3 display height.

I also usually keep one Chrome window logged into a specific Google profile on the side monitor, but I imagine it'll be harder to position a very specific Chrome window like that and not move all the others.

Edit: Ahh, just seen [this note](https://github.com/rxhanson/RectanglePro-Community/discussions/45#discussioncomment-2493448) that there is nop Apple API for this. Nothing new in Ventura either to be able to support this?

**Reply by rxhanson** [reply]
Specific display: yes. This would be configured per app window, not per group.
Specific desktops (spaces): no, and there is nothing new in Ventura for this. Ventura's only addition is Stage Manager, and there's no public API for that, either.

**Comment by davidski** [comment]
I recently discovered this feature and am excited about the possibilities! I don't see a way to have a group arrange the desired applications and *hide all other* active windows. Am I missing a configuration somewhere?

My use case is I'd like to bring up my working applications and hide all other distractions with a single shortcut. Thanks for the help!

**Reply by rxhanson** [reply]
Excellent point; I hadn't thought of that one. I'll add it to the queue.

**Reply by ruaanvds** [reply]
Adding my +1 for this feature.

My workaround at the moment: 

```
osascript -e 'tell application "System Events"
    set visibleApps to every process whose visible is true and name is not "Finder"
    repeat with theApp in visibleApps
        set visible of theApp to false
    end repeat
end tell'

open -g "rectangle-pro://execute-layout?name=Admin"
```

Works okay with a tool like Raycast's [script commands](https://www.raycast.com/blog/getting-started-with-script-commands), but feels clunky.

**Reply by FrancoLab** [reply]
This feature would be amazing. I love Rectangle Pro. I currently use it with multiple desktop layouts, but if there was an easy way to do this without the need to switch desktops, it would be awesome.

**Reply by exegeteio** [reply]
+1 here, too.  At the moment I'm using Spaces to jump between apps for dev work.  But that means manually opening / moving windows to specific spaces.  I'd much rather have a single space with a shortcut keys to hide everything and center/maximize one specific app.

**Reply by asknet** [reply]
+1 from my end. Unfortunately none of the other apps also support this.

**Reply by hybridvision** [reply]
I wonder if [Aerospace](https://github.com/nikitabobko/AeroSpace) could be used in conjunction with Rectangle for this?

For those unfamiliar with Aerospace, it's a tiling window manager with [its own implementation of spaces](https://nikitabobko.github.io/AeroSpace/guide#emulation-of-virtual-workspaces). It's the only one I've been able to stick with for any period of time and I generally like it a lot but there are still times it's a little heavy-handed for me (hence my exploration of other tools to augment it).

The [Aerospace Workspaces](https://nikitabobko.github.io/AeroSpace/guide#emulation-of-virtual-workspaces) are really neat because you can have a heap of them with the ability to easily switch instantly or send an app to any of them via shortcuts. It's so much better than the native Mac spaces implementation.

I imagine most Rectangle users prefer manual control over window sizing and placement and might be annoyed by a window manager but Aerospace can operate in "floating mode", which makes it much less invasive but still gives you all the workspace features such as automatically placing new app windows in specific spaces. For [reference](https://nikitabobko.github.io/AeroSpace/guide#on-window-detected-callback), this is how you enable floating-only behaviour:

```toml
[[on-window-detected]]
check-further-callbacks = true
run = 'layout floating'
```

Alternatively, you can let Aerospace manage all the windows and manually pop specific windows into floating mode when needed. However, then you need some other tooling like Rectangle to move and resize these floating windows.

Bonus tip: if you're on a much larger screen (eg. OLED TV), another neat little app I've used is [Blurred](https://github.com/dwarvesf/blurred), which allows everything outside the focused window to be blacked out. It's great for focusing on a single app when that app would otherwise be uncomfortable fully maximised.

I'm still searching for my perfect setup but hopefully someone else finds this useful 🙂

**Reply by exegeteio** [reply]
@hybridvision Thanks for the suggestion.  I have experimented with using Aerospace and Amethyst along with Rectangle, and I find it painful to have two window managers running at the same time.  I also find tiling window managers to be rough in MacOS as some MacOS windows have a minimum size, or a fixed size, which just doesn't play well with them.  Not a slight against those window managers, more against the apps which enforce those size restrictions.  Rectangle / Rectangle Pro do not have that limitation.

**Reply by hybridvision** [reply]
Hi @exegeteio, I agree and I have similar pain points with rigid MacOS windows. Aerospace is pretty smart and can be configured to remove some these rough edges but my suggestion wasn't so much about having two window managers simultaneously – instead, using Aerospace for the lightweight spaces and something like Rectangle for actual window management. I currently use a mix with Raycast but I like the look of some of Rectangle Pro's features, so I'll be experimenting with that.

**Reply by exegeteio** [reply]
The “[Hide/minimize/quit] Applications not in this layout” feature was added.  It handles everything I need.

**Comment by marepinta** [comment]
Here is how to save a layout:

https://github.com/rxhanson/RectanglePro-Community/discussions/111

You can use URLs to activate a layout, which you could script from within Shortcuts. Here's a terminal command for it.

open -g "rectangle-pro://execute-layout?name=applayout1"
found it!!

You use the name that you gave the layout in Rectangle Pro.

Here is the readme in the community tracking repo where this functionality is mentioned:

https://github.com/rxhanson/RectanglePro-Community#programmatically-triggering-window-sizes--positions

**Reply by marepinta** [reply]
open -g "rectangle-pro://execute-layout?name=ann"

**Comment by D3478** [comment]
Is there anyway to restore or open the apps automatically when clicking on saved layout ?

**Reply by rxhanson** [reply]
You can check the box for `Launch closed/minimized apps` in the configuration popover to open apps, and you can add a saved layout to the menu bar menu in the `Menu & Icon` tab of the Settings window. Does that answer your question?

**Reply by D3478** [reply]
yup ,, thanks

**Comment by michaelaye** [comment]
Is this only available in Pro? Don't see that menu entry in the pop-up menu from the menubar.

**Reply by rxhanson** [reply]
Indeed, this is only available in Rectangle Pro.

**Comment by kerryj89** [comment]
You know how you can apply layout based on monitor connected/disconnected?  Are you able to also re-check for this when the Rectangle app itself launches?

**Reply by rxhanson** [reply]
I don't have anything in at the moment that will do this, but can add it in. Is your expectation that Rectangle Pro keeps track of the connected displays that it last encountered, and when launched will assess which displays were added or removed and apply layouts accordingly?

**Reply by kerryj89** [reply]
Yes, I think so.  Let's say I **don't** have the app running at home, but when I come to work, I would like to start Rectangle app up.  I would like for it to be able to determine which layout to use by monitor(s) connected... thinking about this, I guess it only applies layout based on a singular monitor connect/disconnect event versus "these 2 monitors are connected currently"?  So maybe keeping track or remembering last setup won't help too much now that I think about it unless I'm misunderstanding you and the intent is to diff against it.  Thanks for considering!
